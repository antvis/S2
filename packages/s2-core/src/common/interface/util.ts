export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
