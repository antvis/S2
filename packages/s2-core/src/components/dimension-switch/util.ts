import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
