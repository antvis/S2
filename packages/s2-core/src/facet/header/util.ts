import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
