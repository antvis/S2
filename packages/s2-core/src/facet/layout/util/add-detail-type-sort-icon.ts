import { Group } from '@antv/g-canvas';
import * as _ from 'lodash';
import { GuiIcon } from '../../../common/icons';
import { DEFAULT_PADDING, KEY_LIST_SORT } from '../../../common/constant';
import { SpreadSheet } from '../../..';

// 排序按钮的宽度
export const SORT_ICON_SIZE = 8;

export const getIconType = (key: string, spreadsheet: SpreadSheet) => {
  /*
  currentSortKey 存储的点击的某个field对应的升序还是降序
  currentSortKey = {
   [field]: up/down
  }
   */
  const currentSortKey = spreadsheet.store.get('currentSortKey', {});
  let upIconType = 'SortUp';
  let downIconType = 'SortDown';
  if (currentSortKey && _.has(currentSortKey, key)) {
    // 有配置,当前点击的过的key(某个维度)
    if (_.get(currentSortKey, key) === 'up') {
      // 点击过此维度的up
      upIconType = 'SortUpSelected';
    } else {
      // 点击过此维度的down
      downIconType = 'SortDownSelected';
    }
  }

  return {
    upIconType,
    downIconType,
  };
};

export const renderIcon = (
  parent: Group,
  spreadsheet: SpreadSheet,
  x: number,
  y: number,
  type: string,
  key: string,
  sortKeyVal: string,
  sortType: string,
) => {
  const icon = new GuiIcon({
    type,
    x,
    y,
    width: SORT_ICON_SIZE,
    height: SORT_ICON_SIZE,
  });

  icon.on('click', () => {
    const selected = icon.get('selected');
    let currentSortKey = {};
    if (!selected) {
      currentSortKey = {
        [key]: sortKeyVal,
      };
    }
    spreadsheet.store.set('currentSortKey', currentSortKey);
    icon.set('selected', !selected);
    spreadsheet.emit(KEY_LIST_SORT, {
      sortFieldId: key,
      sortMethod: selected ? '' : sortType,
    });
    spreadsheet.render(false);
  });

  parent.add(icon);
};

export const renderDetailTypeSortIcon = (
  parent: Group,
  spreadsheet: SpreadSheet,
  textX: number,
  textY,
  key,
) => {
  const iconType = getIconType(key, spreadsheet);

  renderIcon(
    parent,
    spreadsheet,
    textX,
    textY - DEFAULT_PADDING * 2,
    iconType.upIconType,
    key,
    'up',
    'ASC',
  );
  renderIcon(
    parent,
    spreadsheet,
    textX,
    textY - DEFAULT_PADDING / 2,
    iconType.downIconType,
    key,
    'down',
    'DESC',
  );
};
