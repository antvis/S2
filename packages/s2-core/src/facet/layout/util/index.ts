export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
