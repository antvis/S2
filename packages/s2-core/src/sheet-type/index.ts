import { PivotSheet } from './pivot-sheet';
import { TableSheet } from './table-sheet';
import { SpreadSheet } from './spread-sheet';

export { PivotSheet, TableSheet, SpreadSheet };
