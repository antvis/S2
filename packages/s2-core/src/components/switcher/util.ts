import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
