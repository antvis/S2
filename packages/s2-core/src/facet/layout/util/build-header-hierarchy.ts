import {
  size,
  includes,
  isEmpty,
  each,
  merge,
  isBoolean,
  remove,
} from 'lodash';
import { Pivot } from '../../../data-set';
import { i18n } from '../../../common/i18n';
import { SpreadsheetFacetCfg } from '../../../common/interface';
import { Hierarchy } from '../hierarchy';
import { Node } from '../node';
import TotalClass from '../total-class';
import { canNodeBeExpanded } from './can-node-be-expanded';
import getDimsCondition from './get-dims-condition-by-node';
import { EXTRA_FIELD } from '../../../common/constant';
import { handleHideNodes } from './handle-hide-nodes';
import { generateId } from './generate-id';

function addTotalsNodes(
  pivot: Pivot,
  field: string,
  fieldValues: (string | TotalClass)[],
  isSubTotals: boolean,
) {
  const totalsConfig = pivot.getTotalsConfig(field);
  const option = isSubTotals
    ? totalsConfig.showSubTotals
    : totalsConfig.showGrandTotals;
  if (option) {
    // eslint-disable-next-line no-nested-ternary
    const func = isSubTotals
      ? totalsConfig.reverseSubLayout
        ? 'unshift'
        : 'push'
      : totalsConfig.reverseLayout
      ? 'unshift'
      : 'push';
    const value = isSubTotals
      ? new TotalClass(totalsConfig.subLabel, true)
      : new TotalClass(totalsConfig.label, false, true);
    fieldValues[func](value);
  }
}

function handleTotals(
  field: string,
  fields: string[],
  index,
  fieldValues: (string | TotalClass)[],
  pivot: Pivot,
) {
  // put all totals nodes in appropriate place
  if (field !== fields[0]) {
    // sub totals from 1 to length - 1
    const previousField = fields[index - 1];
    // NOTE: if current field's only has one child, we don't need sub totals
    if (size(fieldValues) > 1) {
      // [a, b, c] (start from second index, b -> a, c -> b)
      addTotalsNodes(pivot, previousField, fieldValues, true);
    }
  } else if (field === fields[0]) {
    // grand totals in first field level
    // check if there are grand totals in first field
    addTotalsNodes(pivot, fields[0], fieldValues, false);
  }
}

export default function buildHeaderHierarchy(
  pivot: Pivot,
  parent: Node,
  field: string,
  fields: string[],
  cfg: SpreadsheetFacetCfg,
  hierarchy: Hierarchy,
) {
  const index = fields.indexOf(field);
  const key = field;
  const query = getDimsCondition(parent, true);
  const fieldValues: (string | TotalClass)[] = pivot.getDimValues(key, query);

  // when cols/values not exist,we need show table structure with empty body
  if (key === EXTRA_FIELD && includes(cfg.cols, key) && isEmpty(fieldValues)) {
    // add empty measure node
    fieldValues.push('');
  }

  // handle handle nodes by node query
  handleHideNodes(fieldValues, field, cfg);

  // handle totals nodes
  handleTotals(field, fields, index, fieldValues, pivot);
  const { collapsedCols } = cfg;
  // generate all nodes(normal and totals) of this field
  each(fieldValues, (col) => {
    const isTotal = col instanceof TotalClass;
    let value;
    if (isTotal) {
      value = i18n((col as TotalClass).label);
    } else {
      value = col;
    }
    const nodeQuery = merge({}, query, { [field]: value });
    const id = generateId(parent.id, value, cfg);
    if (!id) return;
    const isCollapsed = isBoolean(collapsedCols[id])
      ? collapsedCols[id]
      : false;
    const node = new Node({
      id,
      key,
      value,
      level: index,
      field,
      parent,
      isTotals: isTotal,
      isCollapsed,
      isSubTotals: isTotal ? (col as TotalClass).isSubTotals : false,
      isGrandTotals: isTotal ? (col as TotalClass).isGrandTotals : false,
      hierarchy,
      query: nodeQuery,
      spreadsheet: cfg.spreadsheet,
    });

    if (cfg.hierarchy) {
      const another = cfg.hierarchy(cfg.spreadsheet, node);
      const push = another?.push;
      if (another.nodes.length > 0) {
        // 存在节点，按push 或者 unshift插入
        if (push) {
          hierarchy.pushNode(node);
        }
        each(another.nodes, (v) => {
          hierarchy.pushNode(v);
        });
        if (!push) {
          // new node insert before current
          hierarchy.pushNode(node);
          // adjust the index of current node
          remove(parent.children, (v) => v === node);
          parent.children.push(node);
        }
      } else {
        // 不存在节点，正常插入
        hierarchy.pushNode(node);
      }
    } else {
      // no extra node exist
      hierarchy.pushNode(node);
    }

    if (index > hierarchy.maxLevel) {
      hierarchy.sampleNodesForAllLevels.push(node);
      hierarchy.sampleNodeForLastLevel = node;
      hierarchy.maxLevel = index;
    }

    if (index === fields.length - 1 || isTotal || isCollapsed) {
      node.isLeaf = true;
      hierarchy.pushIndexNode(node);
      node.rowIndex = hierarchy.getIndexNodes().length - 1;
    }
    if (canNodeBeExpanded(index < fields.length - 1, isCollapsed, isTotal)) {
      buildHeaderHierarchy(
        pivot,
        node,
        fields[index + 1],
        fields,
        cfg,
        hierarchy,
      );
    }
  });
}
