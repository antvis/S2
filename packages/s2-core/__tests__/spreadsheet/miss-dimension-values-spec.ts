import { getContainer } from 'tests/util/helpers';
import {
  EMPTY_FIELD_VALUE,
  LayoutWidthType,
  ORIGIN_FIELD,
  type S2DataConfig,
  type S2Options,
} from '@/common';
import { PivotSheet, SpreadSheet } from '@/sheet-type';

const s2Options: S2Options = {
  debug: true,
  width: 600,
  height: 400,
  hierarchyType: 'grid',
  totals: {
    row: {
      showGrandTotals: false,
      showSubTotals: {
        always: false,
      },
      reverseGrandTotalsLayout: false,
      reverseSubTotalsLayout: false,
      subTotalsDimensions: ['first', 'second'],
    },
    col: {
      showGrandTotals: false,
      showSubTotals: false,
      reverseGrandTotalsLayout: false,
      reverseSubTotalsLayout: false,
      subTotalsDimensions: [],
    },
  },
  style: {
    layoutWidthType: LayoutWidthType.Adaptive,
    dataCell: {
      height: 30,
    },
  },
  showDefaultHeaderActionIcon: false,
};

const testDataCfg: S2DataConfig = {
  meta: [
    {
      field: 'first',
      name: '一级维度',
    },
    {
      field: 'second',
      name: '二级维度',
    },
    {
      field: 'third',
      name: '三级维度',
    },
    {
      field: 'number',
      name: '数值',
    },
  ],
  fields: {
    rows: ['first', 'second', 'third'],
    columns: [],
    values: ['number'],
    valueInCols: true,
  },
  data: [
    {
      first: '总计',
      number: 1732771,
    },
    {
      first: '维值-1',
      second: '维值-2',
      third: '维度-3',
      number: 172245,
    },
    {
      first: '维值-1',
      second: '维值-2',
      third: '维度-3',
      number: 12222,
    },
    {
      first: '维值-1',
      second: '维值-3',
      third: '维值-3',
      number: 11111,
    },
    {
      first: '维值-1',
      second: '维值-3',
      third: '维度-3',
      number: 11111,
    },
    {
      first: '维值-1',
      number: 456,
    },
    {
      first: '测试-1',
      second: '测试-2',
      third: '维度-3',
      number: 12,
    },
    {
      first: '测试-1',
      second: '测试-2',
      third: '维度-3',
      number: 4444567,
    },
    {
      first: '测试-1',
      second: '测试-3',
      number: 111233,
    },
    {
      first: '测试-1',
      second: '测试-3',
      number: 785222,
    },
    {
      first: '测试-1',
      second: '测试-4',
      third: '维度-3',
      number: 6455644,
    },
    {
      first: '测试-1',
      second: '测试-4',
      number: 289898,
    },
    {
      first: '测试-1',
      second: '测试-5',
      number: 2222,
    },
    {
      first: '测试-1',
      second: '测试-5',
      third: '维度-3',
      number: 1111,
    },
    {
      first: '测试-1',
      number: 125555,
    },
    {
      first: '测试-6',
      second: '测试-x',
      number: 409090,
    },
    {
      first: '测试-6',
      second: '测试-x',
      number: 111111,
    },
    {
      first: '测试-6',
      second: '测试-7',
      number: 5555,
    },
    {
      first: '测试-6',
      second: '测试-7',
      number: 67878,
    },
    {
      first: '测试-6',
      second: '测试-8',
      number: 53445.464,
    },
    {
      first: '测试-6',
      second: '测试-8',
      number: 456.464,
    },
    {
      first: '测试-6',
      number: 123.416,
    },
  ],
};

describe('Miss Dimension Values Tests', () => {
  let s2: SpreadSheet;

  beforeEach(async () => {
    s2 = new PivotSheet(getContainer(), testDataCfg, s2Options);
    await s2.render();
  });

  test('should get correctly empty dimension values', () => {
    const emptyDimensionValueNode = s2.facet.getRowNodes()[0].children[0];

    expect(emptyDimensionValueNode.value).toEqual(EMPTY_FIELD_VALUE);
    expect(emptyDimensionValueNode.id).toEqual(
      `root[&]总计[&]${EMPTY_FIELD_VALUE}`,
    );
    expect(emptyDimensionValueNode.belongsCell!.getActualText()).toEqual('-');
  });

  test('should get correctly empty dimension values and use custom placeholder text', async () => {
    const placeholder = '*';

    s2.setOptions({
      placeholder,
    });

    await s2.render(false);

    const emptyDimensionValueNode = s2.facet.getRowNodes()[0].children[0];

    expect(emptyDimensionValueNode.belongsCell!.getActualText()).toEqual(
      placeholder,
    );
  });

  test('should generate correct query for empty node', () => {
    const emptyDimensionValueNode1 = s2.facet.getRowNodes()[0];

    expect(emptyDimensionValueNode1.query).toEqual({
      first: '总计',
    });

    const emptyDimensionValueNode2 = s2.facet.getRowNodes()[1];

    expect(emptyDimensionValueNode2.query).toEqual({
      first: '总计',
    });
  });

  test('should get correctly dimension data and ignore empty dimension value', () => {
    const emptyDimensionValueNode = s2.facet.getRowNodes()[0].children[0];

    const data = s2.dataSet.getCellMultiData({
      query: emptyDimensionValueNode.query!,
    });
    const dimensionValues = s2.dataSet.getDimensionValues(
      emptyDimensionValueNode.field,
    );
    const emptyDimensionDataCell = s2.facet.getDataCells()[0];

    expect(emptyDimensionValueNode.query).toEqual(
      emptyDimensionValueNode.parent!.query,
    );
    expect(emptyDimensionDataCell.getMeta().fieldValue).toEqual(1732771);
    expect(data[0][ORIGIN_FIELD]).toMatchInlineSnapshot(`
      Object {
        "first": "总计",
        "number": 1732771,
      }
    `);
    expect(dimensionValues).toMatchInlineSnapshot(`
      Array [
        "维值-2",
        "维值-3",
        "测试-2",
        "测试-3",
        "测试-4",
        "测试-5",
        "测试-x",
        "测试-7",
        "测试-8",
      ]
    `);
  });
});
