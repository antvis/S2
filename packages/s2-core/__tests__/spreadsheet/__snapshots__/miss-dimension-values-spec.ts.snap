// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Miss Dimension Values Tests should replace empty dimension value placeholder when copy formatted data 1`] = `
"一级维度	二级维度	三级维度	数值
总计	-	-	1732771
维值-1	维值-2	维度-3	12222
维值-1	维值-3	维值-3	11111
维值-1	维值-3	维度-3	11111
维值-1	维值-3	小计	-
维值-1	小计		456
测试-1	测试-2	维度-3	4444567
测试-1	测试-3	-	785222
测试-1	测试-4	维度-3	6455644
测试-1	测试-5	维度-3	1111
测试-1	小计		125555
测试-6	测试-x	-	111111
测试-6	测试-7	-	67878
测试-6	测试-8	-	456.464
测试-6	小计		123.416"
`;

exports[`Miss Dimension Values Tests should replace empty dimension value placeholder when copy original data 1`] = `
"first	second	third	number
总计	$$empty_field_value$$	$$empty_field_value$$	1732771
维值-1	维值-2	维度-3	12222
维值-1	维值-3	维值-3	11111
维值-1	维值-3	维度-3	11111
维值-1	维值-3	小计	
维值-1	小计		456
测试-1	测试-2	维度-3	4444567
测试-1	测试-3	$$empty_field_value$$	785222
测试-1	测试-4	维度-3	6455644
测试-1	测试-5	维度-3	1111
测试-1	小计		125555
测试-6	测试-x	$$empty_field_value$$	111111
测试-6	测试-7	$$empty_field_value$$	67878
测试-6	测试-8	$$empty_field_value$$	456.464
测试-6	小计		123.416"
`;
